hugolib/page.go

```diff
@@ -468,13 +468,46 @@ func (p *Page) LinkTitle() string {
 }
 
 func (p *Page) ShouldBuild() bool {
-	if (viper.GetBool("BuildFuture") || p.PublishDate.IsZero() || p.PublishDate.Before(time.Now())) &&
-		(viper.GetBool("BuildExpired") || p.ExpiryDate.IsZero() || p.ExpiryDate.After(time.Now())) {
-		if viper.GetBool("BuildDrafts") || !p.Draft {
-			return true
+	return AssertShouldBuild(viper.GetBool("BuildFuture"), viper.GetBool("BuildExpired"),
+		viper.GetBool("BuildDrafts"), p.Draft, p.PublishDate, p.ExpiryDate)
+}
+
+func AssertShouldBuild(buildFuture bool, buildExpired bool, buildDrafts bool, Draft bool,
+	publishDate time.Time, expiryDate time.Time) bool {
+	if !buildFuture {
+		if IsFuturePage(publishDate) {
+			return false
 		}
 	}
-	return false
+	if !buildExpired {
+		if IsExpiredPage(expiryDate) {
+			return false
+		}
+	}
+	if !(buildDrafts || !Draft) {
+		return false
+	}
+	return true
+}
+
+func IsExpiredPage(expiryDate time.Time) bool {
+	if expiryDate.IsZero() {
```

Comment:
Simplify this function and `IsFuturePage` to just be:

```go
return !expiryDate.IsZero() && !expiryDate.After(time.Now())
```

Or just get rid of these functions and roll that logic back into `AssertShouldBuild`.